client/src/components/ElementEditor/Element.js

```diff
@@ -53,6 +74,72 @@ class Element extends Component {
         captureDraggingState: true,
       });
     }
+    // Check if formSchema state has already been loaded before opening a block
+    // This can happen if there was a validation error on a block after performing a Page save
+    if (this.props.formStateExists) {
+      this.setState({
+        formHasRendered: true
+      });
+    }
+  }
+
+  componentDidUpdate() {
+    if (this.state.justClickedPublishButton && this.state.formHasRendered) {
+      this.setState({
+        justClickedPublishButton: false,
+      });
+      if (this.props.formDirty) {
+        // Save the element first before publishing, which may trigger validation errors
+        this.props.submitForm();
+        this.setState({
+          doPublishElementAfterSave: true,
+        });
+      } else {
+        // Just publish the element straight away without saving first
+        this.setState({
+          doPublishElement: true,
+        });
+      }
+    }
+    if (this.state.doDispatchAddFormChanged) {
+      this.setState({
+        doDispatchAddFormChanged: false,
+      });
+      this.props.dispatchAddFormChanged();
+    }
```

Comment:
```suggestion
    const newState = {};
    if (this.state.justClickedPublishButton && this.state.formHasRendered) {
      newState.justClickedPublishButton = false;
      if (this.props.formDirty) {
        // Save the element first before publishing, which may trigger validation errors
        this.props.submitForm();
        newState.doPublishElementAfterSave = true;
      } else {
        // Just publish the element straight away without saving first
        newState.doPublishElement = true;
      }
    }
    if (this.state.doDispatchAddFormChanged) {
      newState.doDispatchAddFormChanged = false;
      this.props.dispatchAddFormChanged();
    }
    this.setState(newState);
```
We only need to set state once here (unless those function calls require the new state - but I think the actual **setting** of the state happens in the next event cycle anyway?)

Reply:
Upated